### DISCUSSION TITLE
Deduplication in Missa

### DISCUSSION THREAD

**Discussion by Michaelskapel** [DivinumOfficium/divinum-officium#3993] [Q&A]
In my effort to deduplicate the text items in Missa, I have compiled a redirection table that shows all possibilities to redirect a label in a certain code file to a label in a different code file. Almost all redirections (somewhere between 1500 and 2000!) have been implemented in the Nederlands version. Does it make sense to put the tabel (Redirection_Table.xlsx) and a description (Deduplication_in_Divinum_Officium_Missa.docx) on my Google drive and make it available to others?

**Comment by APMarcello3** [comment]
Please do!

**Reply by Michaelskapel** [reply]
Here are the links to **Deduplication_in_Divinum_Officium_Missa.docx**

https://docs.google.com/document/d/14OjoKWnH_Efi15_nDP7Qqg0PK27do6GU/

and to **Redirection_Table.xlsx**

https://docs.google.com/spreadsheets/d/14L5PNvKliQyTJmRPk4H7jAgm3rf20xk7/

The table, which is a work in progress, is without any warranty. I worked very carefully, but errors may still be present. Please double check before you implement a redirection: once the redirection is done, it is not easy to detect an error.

**Comment by mmolenda** [comment]
Hello, I'm not sure if it is directly related to this effort, but recently I spotted numerous refactoring and deduplication commits, mostly in scope of introduction of the Spanish language version by the author @d4h0y0sr. Some of the changes are not clear to me though. Could the author clarify several selected cases for me please?

- Why is Oratio section removed from web/www/missa/English/Sancti/01-01.txt while the parent proper is Sancti/12-25m3 which has a different Oratio? [commit](https://github.com/DivinumOfficium/divinum-officium/commit/8b82895f111bf4ea2eae468190a78a5fc32cc7bb#diff-c52cfc6a53792c8ae2501474c4e9bcdb754fcb31c5239840264993e7bcf300eb)
- Why is Postcommunio removed from web/www/missa/English/Sancti/12-27.txt? [commit](https://github.com/DivinumOfficium/divinum-officium/commit/5a8ff94a58b3be2430f55f9f61774f57f21afb6f#diff-7c0803f2b98ed0bd19ea54d576c30298729a706fbb1f15d7b8bdd1ab3faa8acd)
- Why is Postcommunio removed from web/www/missa/English/Commune/C2c.txt? [commit](https://github.com/DivinumOfficium/divinum-officium/commit/a8028cd5c5fa81fab8e219867df5521709bca1e8#diff-a3868bfeee9e185806eafd08f592cabbcf4919ac48243e7f4848d09350990f12)

**Reply by mmolenda** [reply]
Yet another example: commemoration of St. Maur [removed altogether](https://github.com/DivinumOfficium/divinum-officium/commit/0499a5dbbf088d7293927f906776b7ce9dda3b4f#diff-619f3e452a729eb2451ee4f4bde3ef7a20b72bbe005a0ad8079fa2cbfd70fbb6). No references added in any of subsequent commits.

**Reply by APMarcello3** [reply]
For St. Maur (01-15cc), aa6f762 should at least partially remedy the problem. The headline "Commemoratio ad Laudes tantum: " with the missing saint's name (in this case, Maur) is now restored as well.

**Reply by wolfstain** [reply]
> Hello, I'm not sure if it is directly related to this effort, but recently I spotted numerous refactoring and deduplication commits, mostly in scope of introduction of the Spanish language version by the author @d4h0y0sr. Some of the changes are not clear to me though. Could the author clarify several selected cases for me please?
> 
>     * Why is Oratio section removed from web/www/missa/English/Sancti/01-01.txt while the parent proper is Sancti/12-25m3 which has a different Oratio? [commit](https://github.com/DivinumOfficium/divinum-officium/commit/8b82895f111bf4ea2eae468190a78a5fc32cc7bb#diff-c52cfc6a53792c8ae2501474c4e9bcdb754fcb31c5239840264993e7bcf300eb)
> 
>     * Why is Postcommunio removed from web/www/missa/English/Sancti/12-27.txt? [commit](https://github.com/DivinumOfficium/divinum-officium/commit/5a8ff94a58b3be2430f55f9f61774f57f21afb6f#diff-7c0803f2b98ed0bd19ea54d576c30298729a706fbb1f15d7b8bdd1ab3faa8acd)
> 
>     * Why is Postcommunio removed from web/www/missa/English/Commune/C2c.txt? [commit](https://github.com/DivinumOfficium/divinum-officium/commit/a8028cd5c5fa81fab8e219867df5521709bca1e8#diff-a3868bfeee9e185806eafd08f592cabbcf4919ac48243e7f4848d09350990f12)

English is a dependance of latin, and every other languages depends on english, we are using inheritance. If something is wrong it should be fixed in the latin file.

Examples:

Sancti/01-01.txt  ---> Oratio comes from @Commune/C10b. 
Sancti/12-27.txt  ---> Oratio comes from @Commune/C5.
etc.

If someone puts any text in English, derived languages wont display the correct references. English does not need to have any reference because by default it will inheritance and use the reference of the latin file, thats why I remove english blocks, they are redundant.

**Reply by mmolenda** [reply]
OK, I got it, thank you for clarification. So when Latin/Sancti/01-01.txt has this section:

```
[Oratio]
@Commune/C10b
```

then there is no need to define Oratio in English/Sancti/01-01.txt. It only needs to contain English bits specific for this proper.
Is it something that was introduced recently, or just the vernacular files got cluttered with redundant sections over time?

**Reply by d4h0y0sr** [reply]
Exactly. If you put Oratio in English will be redundant because it already exist the text in C10b, and it will harm the translations that rely on the latin references because in other languages will be displayed the English text instead of the vernacular reference. 

I dont know if the references system are something new, I have to learn by my self and with help from these forums because documentation does not clarify this. I would say vernaculars are cluttered with redundant sections because it was straightforward to copy and paste on the files, so even if you dont know programming you can just put whatever text in the files and it works. The problem is when you try to do translations by yourself, like i had been doing in spanish, or like Michaëlskapel has been doing with nederlands, you find out this is a very repetitive and tedious task, and is difficult to maintain because if some typo is present in one place you need to fix it in a lot of places.

**Reply by Michaelskapel** [reply]
A few remarks from my side.

- When I started to work on the Netherlands version, I took the Latin version as my template and simply changed the Latin text into Nederlands. At that time, the Latin version also had many duplications and almost no redirections.

- I knew that the English version played a special role, but was not expecting that differences between Latin and English would lead to undesired results in other vernaculars. Anyway, if I noticed such things happening, they could be repaired by adding items also to the Nederlands version in case they were present in the English version but not in the Latin one. I already complained to father Marcello about this. Does one really have to monitor what is happening (in the future!) with the English version and repair things in one's own vernacular? That would be ridiculous!

- While filling the Nederlands version and seeing all the duplications, I started to construct my [redirection table](https://docs.google.com/spreadsheets/d/14L5PNvKliQyTJmRPk4H7jAgm3rf20xk7/). The table is now available for everyone's use. Even if the redirections will not be implemented in the vernaculars that are already present, the table shows you where identical texts appear, so that you know all the places that need attention if you want to change something.

- I take it that you have seen my [deduplication document](https://docs.google.com/document/d/14OjoKWnH_Efi15_nDP7Qqg0PK27do6GU/);  I have added a few things, mainly in italics. Note also that I have been working on the redirection table after I mentioned it when this discussion started. The table now also shows the redirection possibilities between the three missing text items that do not have their own label: Graduale, Alleluia, and Tractus. Whether or not the labels [Graduale] [Tractus] [GradualeP] etc. will be redesigned to allow deduplication of Graduale, Alleluia, and Tractus, the table will still be a big help in maintenance, because it shows which texts are identical. 

- In the ideal case, the Latin version should have only unique texts and all the redirections. New vernaculars can then be built on the Latin version very easily: one only has to replace the Latin text by its own vernacular and one would not have to worry about redirections, because they are already there. **You can imagine how much time can be saved in creating a version in a new vernacular!**

**Reply by mmolenda** [reply]
@d4h0y0sr, @Michaelskapel thank you very much for your input and clarification on this.

> The problem is when you try to do translations by yourself, like i had been doing in spanish, or like Michaëlskapel has been doing with nederlands, you find out this is a very repetitive and tedious task

I'm the author of Polish translation for the mass part so I understand the pain. I didn't go much into the details of the references though, I just used them here and there, mostly repeating references from Latin. I appreciate your deduplication effort.

**Reply by d4h0y0sr** [reply]
@mmolenda you should not repeat references from Latin, because thats the whole reason of inheritance. You just need to remove the text from english, it does not affect english, because english language will inherite the reference and resolved, and also polish will do the same.

**Example of english roadblocking the reference flow:**

Latin Reference in section --> English (redundant section with text) :: display texts--> Vernacular (No section) :: algorithm uses English texts

**After you remove the redundant english section:**

Latin Reference in section --> English (No section)  :: algorithm use Latin Reference --> Vernacular (No section) :: algorithm uses Latin reference.

----------------

I wish the algorithm check first if the latin has a reference before choosing to display the english text, that would make life easier.

**Reply by Michaelskapel** [reply]
Personally, I do not feel authorized to work on the English version; I restrict myself to Nederlands and to obvious typos in Latin. 
As long as we keep the system as it is now (Latin --> English --> vernacular), we may need a supervisor whose task it is to keep the English version in sync with the Latin one.

**Reply by Michaelskapel** [reply]
From the recent discussion, I have learned a lot about the Divinum Officium's label system. I want to check with you if my understanding of the system is correct.

1. There are four folders - Commune, Ordo, Sancti, and Tempora - which contain the code files.
2. The code files in the Latin version contain three types of labels, each with their own type of content: (i) text-labels with a (unique) text in Latin (ideally without duplications), (ii) reference-labels with a reference  to another label (which may then contain text or another reference), and (iii) system-labels, such as [Rank] and [Rule], which can better be left unchanged in a vernacular version, except for obvious adaptations like the local names of Saints.
3. It is obvious that a vernacular version shall have the four folders and the location of the code files in sync with the Latin version.
4. The Latin version is leading, and any vernacular version inherits the label system from the Latin version.
5. A vernacular version has the (unique) text-labels in its own language; ideally, this is the only thing that needs to be provided to make a certain vernacular available in Divinum Officium.
6. Since a vernacular version inherits all the references from the leading version, there is - in principal - no need to keep reference-labels in the vernacular version; if the need of a reference-label still arises, such a label contains either (i) text or (ii) a reference to a label that contains this text.
7. Given the special status of the English version - which may take over the leading position from the Latin version - its (unique) text-labels shall be in complete sync with the ones in the Latin version. 
8. The English version is discouraged to include labels that are not present in the Latin version; if such labels cannot be avoided, vernacular versions are forced to also keep such labels (with the appropriate content).

Comments are appreciated.

**Reply by d4h0y0sr** [reply]
1. Yes
2. Yes
3. Yes
4. Yes, in this case vernacular will inherit first from english then if empty from latin.
5. I think, i have never tried to add a new vernacular language from zero.
6. There should not be an scenario where a vernacular has reference but the latin does not has, that usually happens because english is obstructing the inheritance with unique text, the fix require to remove unique text from english, not to add reference label. Also notice we should avoid a reference label to another reference label, the reference label should be always to a text. Reference label that points to another reference label is known as "reference nesting" which is an issue and could generate problems.

7-8. I think so, i mean in terms of unique text (not references), every vernacular has to have the unique text, english does not matter in this scenario, as you have to write down the vernacular text either way. English just matters when you want to use references labels automatically from latin files without writting again these reference labels in vernacular files.

**Reply by Michaelskapel** [reply]
@d4h0y0sr : Thanks for you comments! A few remarks from my side:

4. I take it that by '_in this case_' you actually mean '_in the present case_'.

5. I did it for Nederlands! I took the Latin version as my template, but noticed that there were many text duplications; this was the reason for me to start constructing the redirection table (first to Commune and later to Tempora and Sancti). Having a completely deduplicated Latin version would make things much easier and less time-consuming (!) for new vernaculars.
The Nederlands version is pretty unique at the moment (except for the combined labels like [Graduale], [Tractus], [GradualeP], etc.), but I had to keep reference-labels where the Latin version had a duplicated text-label. I do hope that the Latin version will be deduplicated soon, after which I can then remove all these no longer necessary reference-labels.

6. Note that the need for a reference-label still arises in the case of a reference to those Commune's Oratio, Secreta, and Postcommunio where '**_N._**' or '**_N. et N._**' has to be substituted by the proper vernacular name(s). It may be possible to solve this by defining the substitution in the [Rule] and adapt the program to take it from there, but I can live with the present situation.

7-8. You are right. 
Since I did not feel authorized to touch the English version, I had to add the labels that were necessary to overcome the disturbed inheritance by the English version. But we really have to find a better solution: the program should be adapted such that the English version will never disturb the inheritance from the Latin version.

**Reply by d4h0y0sr** [reply]
4. In this case = any scenario where we expect vernacular to inheritance. Vernacular can not inheritance from latin if there is a section in english.
5. I suggest any translator to help in the Duplication efforts, there is no one in charge to do this task. If you have the missal, which is always the reference, you can be sure to fix the duplication taking into account to respect the accents and punctuation signs thats it, there are other developers that can review your changes in case you make mistakes. In that way, every "translator" would help not only to maintaining a language but also to keep latin in order. 
6.  You are totally right i forgot that scenario.
7.-8. Given present software architecture of the website, i would say every translator is authorized to touch english just to remove inheritance obstruction, that benefits not only your vernacular language but all vernacular languages. Also its easier to catch mistakes in english as i would say is the most used language in the website.

**Reply by FAJ-Munich** [reply]
Maybe a few comments on the vernacular issue and your discussion which seems to be more pressing for the Missale than the Breviary at the moment:

1. Indeed English should not have any texts wherever a reference is used in Latin. However, pruning is harder done than said as – over time – diverging translations have cluttered in the database and the only authoritative source for decision making on the English, ultimately, is Fr. Marcello. 
2. In general, the missa database is not as well kept than the horas unfortunately. Long-term, I hope the database can be merged such that every office/mass has its designated file and there are no divergences in `[Rank]` and Commemorations etc.
3. Within the `[Rank]` hash in the vernaculars, feel free to add your translations of the Feast and don't worry about anything past the first `;;` group. It's ignored and taken from the Latin only.
4. Technically, there would be a hack to get a Language to be stacked on Latin (or any other vernacular) only: you would need to name it, e.g., `Português/Latin-Portugues` instead in the `missa.dialog` configuration. But be aware that, then, nothing is inherited from English anymore. 
5. Whether a "check Latin for @-reference first" mechanism can easily be developed remains to be analysed.
